File: src/main/python/Advent2023/day_02.py
"""
https://adventofcode.com/2023/day/2
--- Day 2: Cube Conundrum ---
"""
import re
from collections import Counter
from math import prod
from typing import Dict, List

from src.main.python.utilities import LOGGER, timing

GAME_REG = re.compile(r'Game (\d+): ')
CUBE_REG = re.compile(r'(\d+) (blue|green|red)')


def parse(games: List[str]) -> Dict[int, Counter]:
    """
    parse the lines and put them into a dictionary of game number
    to Counter objects
    :param games: line in the format of Game <n>: <n> <color>;
    :return: dictionary of games and Counters
    """
    d = dict()
    for game in games:
        d[int(GAME_REG.match(game).group(1))] = Counter({
            x[1]: x[0] for x in sorted(
                [(int(x[0]), x[1]) for x in CUBE_REG.findall(game)],
            )
        })
    return d


@timing
def part_one(games) -> int:
    possible_game = Counter({'red': 13, 'green': 14, 'blue': 15})
    return sum(k for k, v in parse(games).items() if len(possible_game - v) == 3)


@timing
def part_two(games) -> int:
    return sum(prod(x.values()) for x in parse(games).values())


if __name__ == '__main__':
    with open('../../resources/Advent2023/Day02.txt', 'r') as f:
        input_ = f.readlines()
    LOGGER.info(part_one(input_))
    LOGGER.info(part_two(input_))

File: src/main/scala/Advent2023/Day02.scala
/*
https://adventofcode.com/2023/day/2
--- Day 2: Cube Conundrum ---
*/

package Advent2023

import scala.util.matching.Regex
import utilities.Timing

object Day02 {

  val gamePattern: Regex = "Game (\\d+): ".r
  val cubePattern: Regex = "(\\d+) (blue|green|red)".r

  private def gameMap(line: String): collection.mutable.Map[String, Int] =
    val ms = collection.mutable.Map("index" -> 0, "blue" -> 0, "green" -> 0, "red" -> 0)
    ms("index") = gamePattern.findFirstMatchIn(line) match { case Some(m) => m.group(1).toInt }

    for (cubes <- cubePattern.findAllMatchIn(line)) {
      if (ms(cubes.group(2)) < cubes.group(1).toInt) {
        ms(cubes.group(2)) = cubes.group(1).toInt
      }
    }
    ms

  private def parse(input: String): Array[Map[String, Int]] =
    val games = input.split("\n")
    for (game <- games) yield gameMap(game).toMap

  def part1(input: String): Int = {
    parse(input).foldLeft(0) { (sum, game) =>
      if (
          game("blue") <= 14 &&
          game("green") <= 13 &&
          game("red") <= 12
      ) { sum + game("index") }
      else sum
    }
  }

  def part2(input: String): Int = {
    parse(input).foldLeft(0) { (power, game) =>
      power + (game("blue") * game("green") * game("red"))
    }
  }

  def main(args: Array[String]): Unit = {
    val data: String = {
      io.Source.fromResource("Advent2023/Day02.txt").mkString
    }
    println(Timing.timing(part1(data)))  // 2331  // 0.056447208 s
    println(Timing.timing(part2(data)))  // 71585 // 0.00635764 s
  }
}